File: 0799-champagne-tower/0799-champagne-tower.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        # champ = [[0.0 for _ in range(101)] for _ in range(101)]
        curr = [ 0.0 for _ in range(101)]
        next_row = [ 0.0 for _ in range(101)]

        # print(len(ans),len(ans[0]))
        curr[0]=poured
        for i in range(query_row+1):
            for j in range(0,i+1):
                glass = min(1,curr[j])
                rem = curr[j]-glass; 
                curr[j]=glass
                next_row[j]+=rem/2
                if j+1<len(next_row):
                    next_row[j+1]=rem/2 
            if i==query_row: 
                return curr[query_glass]
            curr=next_row
            next_row = [0]*101

        return -1

File: 0799-champagne-tower/0799-champagne-tower.cpp
class Solution {
public:
    double champagneTower(int poured, int query_row, int query_glass) {
        // vector<vector<double>> champ(101,vector<double>(101,0.0));
        vector<double> curr(101,0.0);
        vector<double> next(101,0.0);
        curr[0]=poured;
        for(int i=0;i<=query_row;i++){
            for(int j=0;j<=i;j++){
                double glass =min(1.0,curr[j]);
                double rem = curr[j]-glass;
                curr[j]=glass;
                double split = rem/2;
                next[j]+=max(0.0,split);
                if((j+1)<next.size()) next[j+1]+=max(0.0,split);
            }
            if(i==query_row) return curr[query_glass];
            curr=next;
            next=vector<double>(101,0.0);
        }
        return curr[query_glass];

    }
};